### PR TITLE
server: add an api for turning on metrics from local

### PR DESCRIPTION
**Pull request by nicks**
Hello @landism, @maiamcc,

Please review the following commits I made in branch nicks/localmetrics5:

1e84937932f5633abc503ac17d0ab128e059634c (2020-11-30 13:30:29 -0500)
server: add an api for turning on metrics from local

Code review reminders, by giving a LGTM you attest that:

* Commits are adequately tested
* Code is easy to understand and conforms to style guides
* Incomplete code is marked with TODOs
* Code is suitably instrumented with logging and metrics

### CODE REVIEW

**Comment by nicks** [comment]
I feel like i've written this combination of code enough times in tilt that it was worth trying to abstract it out a bit more

File: internal/user/api.go [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+// 1) other kinds of user preferences (token? analytics opt-in/opt-out?)
+// 2) server-based preferences
+type Prefs struct {
+	// The kind of metrics stack the user is talking to.
+	MetricsMode model.MetricsMode `json:"metricsMode,omitempty" yaml:"metricsMode,omitempty"`
+}
+
+// Read/write metrics setting from a store.
+// Inspired loosely by https://pkg.go.dev/k8s.io/client-go/kubernetes/typed/core/v1#PodInterface
+type PrefsInterface interface {
```
Comment (maiamcc) [inline internal/user/api.go:15]: Just a naming nit, FFTI: isn't naming an interface `PrefsInterface` redundant? I'd just do `Prefs` and call the struct `prefs`, or call the interface `Preferences`/the struct `Prefs`, or something.

File: internal/user/file.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	"gopkg.in/yaml.v3"
+)
+
+const userPrefsFileName = "tilt_user_prefs.yaml"
+
+type filePrefs struct {
+	dir *dirs.TiltDevDir
+}
+
+func NewFilePrefsInterface(dir *dirs.TiltDevDir) *filePrefs {
```
Comment (maiamcc) [inline internal/user/file.go:17]: feels even weirder to call this `New...Interface` when the thing it's returning is just a struct. `NewFilePrefs`?

File: internal/user/fake.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package user
+
+func NewFakePrefsInterface() *FakePrefsInterface {
```
Comment (maiamcc) [inline internal/user/fake.go:3]: and likewise `FakePrefsInterface` should just be `FakePrefs`

File: internal/hud/server/server.go [hunk 4]
```diff
@@ -68,20 +71,23 @@ func ProvideHeadsUpServer(
...
 		a:        analytics,
+		metrics:  metrics,
 		uploader: uploader,
 	}
 
 	r.HandleFunc("/api/view", s.ViewJSON)
 	r.HandleFunc("/api/dump/engine", s.DumpEngineJSON)
 	r.HandleFunc("/api/analytics", s.HandleAnalytics)
 	r.HandleFunc("/api/analytics_opt", s.HandleAnalyticsOpt)
+	r.HandleFunc("/api/metrics_opt", s.HandleMetricsOpt)
```
Comment (maiamcc) [inline internal/hud/server/server.go:90]: somewhere (idk if in `model` or where we define `MetricsOptAction` vs. `AnalyticsOptAction`) it'd be nice to have a comment differentiating Metrics from Analytics.

**Review by maiamcc — APPROVED** [review]
lgtm!

File: internal/user/api.go (hunk 1 above)
Comment (nicks) [reply internal/user/api.go:15]: See the comment 1 line above this, this is taking the Kubernetes naming convention of [ResourceName]Interface.

I'm not a big fan of the convention, but think the benefits of copying another convention outweighs the costs

File: internal/user/fake.go (hunk 3 above)
Comment (nicks) [reply internal/user/fake.go:3]: done

File: internal/user/file.go (hunk 2 above)
Comment (nicks) [reply internal/user/file.go:17]: done

File: internal/hud/server/server.go (hunk 4 above)
Comment (nicks) [reply internal/hud/server/server.go:90]: done

File: internal/user/api.go (hunk 1 above)
Comment (maiamcc) [reply internal/user/api.go:15]: womp womp. Cool!

**Pull request merged**